src/components/LoginForm/LoginForm.tsx

```diff
@@ -0,0 +1,168 @@
+        return 'invalide password';
+    }
+}
+
+function LoginForm() {
+    const { useAppDispatch, useAppSelector } = useReduxStore();
+    const dispatch = useAppDispatch();
+    const navigate = useNavigate();
+    const location = useLocation();
+    const previousLocation = location.state;
+
+    const {
+        value: emailValue,
+        isShowError: emailIsShowError,
+        isInvalid: emailIsInvalid,
+        errorMessage: emailErrorMessage,
+        onChange: emailOnChange,
+        onBlure: emailOnBlure,
+        setErrorMessage: emailSetErrorMessage,
+        setDidTouch: emailSetDidTouch,
+    } = useValidateInput('', emailValidation);
+
+    const {
+        value: passwordValue,
+        isShowError: passwordIsShowError,
+        isInvalid: passwordIsInvalid,
+        errorMessage: passwordErrorMessage,
+        onChange: passwordOnChange,
+        onBlure: passwordOnBlure,
+        setErrorMessage: passwordSetErrorMessage,
+        setDidTouch: passwordSetDidTouch,
+    } = useValidateInput('', passwordValidation);
+
+    const errorMessages = useAppSelector(errorMessagesSelector);
+    const isError = useAppSelector(isErrorSelector);
+    const isLoading = useAppSelector(isLoadingSelector);
+
+    let emaliErrorText: string | undefined;
+    let passwordErrorText: string | undefined;
+
+    if (Array.isArray(errorMessages)) {
+        emaliErrorText = errorMessages?.find((mgs: string) => mgs.includes('email'));
+        passwordErrorText = errorMessages?.find((mgs: string) => mgs.includes('password'));
+    }
+
+    useEffect(() => {
+        if (emaliErrorText) {
+            emailSetDidTouch(true);
+            emailSetErrorMessage(emaliErrorText);
+        }
+
+        if (passwordErrorText) {
+            passwordSetDidTouch(true);
+            passwordSetErrorMessage(passwordErrorText);
+        }
+    }, [emailSetDidTouch, passwordSetDidTouch, emailSetErrorMessage, passwordSetErrorMessage, emaliErrorText, passwordErrorText]);
+
+    const accessToken = getAccessToken();
+
+    useLayoutEffect(() => {
+        if (accessToken) {
+            if (previousLocation) {
+                navigate(previousLocation);
+            } else {
+                navigate(`/`);
+            }
+        }
+
+        return () => {
+            if (accessToken) {
+                if (previousLocation) {
+                    navigate(previousLocation);
+                } else {
+                    navigate(`/`);
+                }
+            }
+        };
+    }, [accessToken, navigate, isLoading, previousLocation]);
+
+    function onSubmit(event: React.FormEvent<HTMLFormElement>) {
+        event.preventDefault();
+        const form = event.target;
+        const fd = new FormData(form as HTMLFormElement);
+
+        const email = fd.get('email') as LoginData['email'] | null;
+        const password = fd.get('password') as LoginData['password'] | null;
+
+        let data: LoginData = {
+            email: '',
```

Comment:
use enum constant to, as a key
```suggestion
            enum LoginFormField {
              Email = "email",
              Password = "password"
            }
            
            
            data = {
             [LoginFormField.Email] : "",
             ...etc
            }
```

use this enum all over the form